phoenix-core/src/main/java/org/apache/phoenix/compile/ScanRanges.java

```diff
@@ -694,4 +694,127 @@ public TimeRange getRowTimestampRange() {
+                int maxKeyLength = SchemaUtil.getMaxKeyLength(schema, ranges);
+                byte[] keyBuffer = new byte[maxKeyLength];
+
+                // Counting Loop
+                int count = 0;
+                while (count < rangesPermutations) {
+                    byte[] lowerBound;
+                    byte[] upperBound;
+
+                    // Note ScanUtil.setKey internally handles the upper/lower exclusive from a Scan
+                    // point of view. It would be good to break it out in the future for rowkey
+                    // construction vs ScanKey construction To handle differences between
+                    // hbase 2 and hbase 1 scan boundaries
+                    int result = ScanUtil.setKey(schema, ranges, slotSpans, position, KeyRange.Bound.LOWER, keyBuffer,
+                            0, 0, slotSpans.length);
+
+                    if (result < 0) {
+                        // unbound
+                        lowerBound = KeyRange.UNBOUND;
+                    } else {
+                        lowerBound = Arrays.copyOf(keyBuffer, result);
+                    }
+
+                    result = ScanUtil.setKey(schema, ranges, slotSpans, position, KeyRange.Bound.UPPER, keyBuffer, 0, 0,
+                            slotSpans.length);
+
+                    if (result < 0) {
+                        // unbound
+                        upperBound = KeyRange.UNBOUND;
+                    } else {
+                        upperBound = Arrays.copyOf(keyBuffer, result);
+                    }
+
+                    // This is already considered inside of ScanUtil.setKey
+                    // we may want to refactor to pull those out.
+                    boolean lowerInclusive = true;
+                    boolean upperInclusive = true;
+                    // If we are not a single key
+                    if (!Arrays.equals(lowerBound, upperBound)) {
+                        upperInclusive = false;
+                    }
+
+                    // Don't send a null range, send an empty range.
+                    if (lowerBound.length == 0 && upperBound.length == 0) {
+                        lowerInclusive = false;
+                        upperInclusive = false;
+                    }
+
+                    KeyRange keyRange = KeyRange.getKeyRange(lowerBound, lowerInclusive, upperBound, upperInclusive);
+                    expandedRanges.add(keyRange);
+
+                    // update position
+                    int i;
+                    for (i = 0; i < position.length; i++) {
+
+                        if (position[i] < ranges.get(i).size() - 1) {
+                            position[i]++;
+                            break;
+                        } else {
+                            position[i] = 0;
+                        }
+                    }
+
+                    if (i == position.length) {
+                        break;
+                    }
+                    rangesPermutations++;
```

Comment:
For the above logic, I really suspect (didn't look at code in details) it should be as follows if we want the returned range list is sorted in ascending order. Could you double check? I might be wrong. (Shall we use Bitmap operations to replace the for loop?)

int count = 0;
while (count < rangesPermutations) {
    ...
	
	for ( ; i < position.length; i++) {
	    int x = position.length - 1 - i;
		if (position[x] < ranges.get(x).size - 1) {
			position[x]++;
			break;
		} else {
			position[x] = 0;
		}
	}

	if (i == position.length) {
		break;
	}

	count++;
}

Reply:
You are correct that doing this from the trailing edge will provide the order constant easier as opposed to resorting.  I'm not sure what you intend by bitmap operations as we are not a strict bitmap.  Can you elaborate?